core/src/main/scala/chisel3/domain/Type.scala

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.domain
+
+import chisel3.{fromIntToLiteral, Data, Element, Printable, UInt, UnknownWidth, Width}
+import chisel3.experimental.SourceInfo
+import chisel3.internal.{throwException, Builder}
+
+/** A [[Data]] that is used to communicate information of a specific domain
+  * kind.
+  */
+final class Type private (val domain: Domain) extends Element { self =>
```

Comment:
Does a `domain.Type` really need to be an Element or Data at all?

Reply:
They do need to be ports, so I think it would either be changing the type that `IO` accepts or adding a new IO creation method just for `domain.Type`. Either of these is doable.

The more critical thing would be that I expect libraries that assume everything in Chisel can be shoved into a `Wire`, like diplomacy, will want to put `domain.Type` into a wire. We might be able to avoid this with changes to libraries so that they don't use a single `Wire`. (This is related to what I was talking to you offline about assuming everything can go into a `Bundle` is making things hard.) WDYT?

Reply:
Ah I see, then I think `Element` is right. If these need to be ports and wired through then I think they run into the same needs as other things like `Properties`. As in the larger conversation about `Bundles`, maybe there's a better way out there but there has to be some way to compose these things together and `Bundle` (well `Record`) remains the only existing avenue.